datafusion/sqllogictest/test_files/nullif.slt

```diff
@@ -97,11 +97,35 @@ SELECT NULLIF(1, 3);
 ----
 1
 
-query I
+query T
 SELECT NULLIF(NULL, NULL);
 ----
 NULL
 
+query R
+select nullif(1, 1.2);
+----
+1
+
+query R
+select nullif(1.0, 2);
+----
+1
+
+query error DataFusion error: Error during planning: Internal error: Failed to match any signature, errors: Error during planning: The signature expected NativeType::String but received NativeType::Int64
+select nullif(2, 'a');
+
+
+query T
+select nullif('2', '3');
+----
+2
+
+# TODO: support numeric string
+# This query success in Postgres and DuckDB
+query error DataFusion error: Error during planning: Internal error: Failed to match any signature, errors: Error during planning: The signature expected NativeType::String but received NativeType::Int64
+select nullif(2, '1');
```

Comment:
This also used to work. 
query T
select nullif(2, '1');
----
2

Interesting that the type is text vs number but still, it did work.

Reply:
This is quite a hard issue to fix.

This query pass in `main` too but it should not.
```
query T
select nullif('1'::varchar, 2);
----
1
```

The change did have regression on `nullif(2, '1')`, but also fix `nullif('1'::varchar, 2)`

Reply:
Similar issue in https://github.com/apache/datafusion/pull/13240#discussion_r1829086791

Reply:
I don't have strong opinions against this PR because it fixes `nullif('1'::varchar, 2)` and refines the `TypeSignature`. We could file a ticket for it and address it in a follow-up.

Reply:
Tracked in https://github.com/apache/datafusion/issues/13285

Reply:
Duckdb can run this query
```
D select nullif('2'::varchar, 2);
┌─────────────────────────────────┐
│ nullif(CAST('2' AS VARCHAR), 2) │
│             varchar             │
├─────────────────────────────────┤
│                                 │
└─────────────────────────────────┘
```

😕

Reply:
To me what is important is not strictly what other systems support - those may/could be a guide to what datafusion will support - but whether the provided arguments to a signature can be losslessly converted to what the signature accepts and whether it logically makes sense to do so. 

I personally would rather be lenient for what is accepted and do casting/coercion as required than to be strict and push the onus onto the user to do that. That's just me though, I don't know if that is the general consensus of the community. Perhaps we should file a discussion ticket with the options and decide?

Reply:
I just double checked on this branch  and it seems good to me (supports things reasonably) 

```sql
     Running `target/debug/datafusion-cli`
DataFusion CLI v43.0.0
> select nullif(2, '1');
+----------------------------+
| nullif(Int64(2),Utf8("1")) |
+----------------------------+
| 2                          |
+----------------------------+
1 row(s) fetched.
Elapsed 0.027 seconds.

> select nullif('1'::varchar, 2);
+----------------------------+
| nullif(Utf8("1"),Int64(2)) |
+----------------------------+
| 1                          |
+----------------------------+
1 row(s) fetched.
Elapsed 0.007 seconds.
```